### DISCUSSION TITLE
Preventing default behaviour for tab key press or overriding it

### DISCUSSION THREAD

**Discussion by balint-kada** [singerdmx/flutter-quill#1941] [General]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/singerdmx/flutter-quill/issues)

### Use case

I would want to implement a way, so tab key can work the same way as in a default text field:
- When pressing tab, text editor loses focus and goes on to the next focus
- In flutter-quill, right now if you press tab, a tab char gets placed into the text

Right now i've tried handling the tab keyevent before it gets handled in the QuillKeyboardListener, but haven't found a way to make it work.

Is there a way to make this work without implementing an override inside flutter-quill?

### Proposal

I would want flutter quill to have an onKeyEvent override, or something similar to quillJs's keyboard module, where you can use keyBindings: https://quilljs.com/docs/modules/keyboard

**Comment by balint-kada** [comment]
Found solution for key presses, with customActions and customShortcuts, but even these don't work for Tab key, they work for every other key.

Does anyone have a workaround for the tab key specificly, or a reason why it works like this here?

**Comment by balint-kada** [comment]
@singerdmx i saw you made it to a discussion, any thoughts?

**Reply by gabrielebrignoli** [reply]
```dart
KeyEventResult _onKeyEvent(node, KeyEvent event) {
    // Don't handle key if there is a meta key pressed.
    if (HardwareKeyboard.instance.isAltPressed ||
        HardwareKeyboard.instance.isControlPressed ||
        HardwareKeyboard.instance.isMetaPressed) {
      return KeyEventResult.ignored;
    }

    if (event is! KeyDownEvent) {
      return KeyEventResult.ignored;
    }
    // Handle indenting blocks when pressing the tab key.
    //if (event.logicalKey == LogicalKeyboardKey.tab) {
    //  return _handleTabKey(event);
    //}

    // Don't handle key if there is an active selection.
    if (controller.selection.baseOffset != controller.selection.extentOffset) {
      return KeyEventResult.ignored;
    }

    // Handle inserting lists when space is pressed following
    // a list initiating phrase.
    if (event.logicalKey == LogicalKeyboardKey.space) {
      return _handleSpaceKey(event);
    }

    return KeyEventResult.ignored;
  }
```

**Reply by balint-kada** [reply]
where would you put this? i tried with something like this with a keyboardlistener wrapping quill editor, it didn't work for me before

**Reply by balint-kada** [reply]
There is a KeyboardListener which prevents handling of key presses inside QuillEditor: 
![Screenshot 2024-06-27 at 13 07 13](https://github.com/singerdmx/flutter-quill/assets/168718905/38111e1c-0671-4cb7-b529-33ad7a816778)
Is this really necessary? Can't we do something else?
